src/frontend/XAPI/XAPIStatement.ts

```diff
@@ -0,0 +1,327 @@
+    const time = truncateDecimalDigits(this.duration);
+    this.addEndSegment(time);
+
+    const data: CompletedDataPlayload = {
+      context: {
+        extensions: {
+          [ContextExtensionsDefintion.sessionId]: this.sessionId,
+        },
+      },
+      result: {
+        completion: true,
+        duration: Interval.fromDateTimes(this.startedAt!, DateTime.utc())
+          .toDuration('milliseconds')
+          .toISO(),
+        extensions: {
+          [ResultExtensionsDefinition.time]: time,
+          [ResultExtensionsDefinition.progress]: 1,
+          [ResultExtensionsDefinition.playedSegment]: this.getPlayedSegment(),
+        },
+      },
+      verb: {
+        display: {
+          'en-US': 'completed',
+        },
+        id: VerbDefinition.completed,
+      },
+    };
+
+    if (contextExtensions.completionTreshold) {
+      data.context!.extensions[ContextExtensionsDefintion.completionTreshold] =
+        contextExtensions.completionTreshold;
+    }
+
+    this.send(data);
+  }
+
+  terminated(
+    contextExtensions: TerminatedContextExtensions,
+    resultExtensions: TerminatedResultExtensions,
+  ): void {
+    const time = truncateDecimalDigits(resultExtensions.time);
+    this.addEndSegment(time);
+
+    const data: DataPayload = {
+      context: {
+        extensions: {
+          [ContextExtensionsDefintion.sessionId]: this.sessionId,
+        },
+      },
+      result: {
+        extensions: {
+          [ResultExtensionsDefinition.time]: time,
+          [ResultExtensionsDefinition.progress]: truncateDecimalDigits(
+            resultExtensions.time / this.duration,
+          ),
+          [ResultExtensionsDefinition.playedSegment]: this.getPlayedSegment(),
+        },
+      },
+      verb: {
+        display: {
+          'en-US': 'terminated',
+        },
+        id: VerbDefinition.terminated,
+      },
+    };
+
+    if (contextExtensions.completionTreshold) {
+      data.context!.extensions[ContextExtensionsDefintion.completionTreshold] =
+        contextExtensions.completionTreshold;
+    }
+
+    this.send(data);
+  }
+
+  interacted(contextExtensions: InteractedContextExtensions): void {
+    // find a way to remove this undefined type. There is no undefined value
+    const extensions: {
+      [key: string]: string | boolean | number | undefined;
+    } = {
+      [ContextExtensionsDefintion.sessionId]: this.sessionId,
+    };
+    for (const key of Object.keys(contextExtensions)) {
+      extensions[
+        ContextExtensionsDefintion[key as keyof InteractedContextExtensions]
+      ] =
+        contextExtensions[key as keyof InteractedContextExtensions];
+    }
+
+    const data: DataPayload = {
+      context: {
+        extensions,
+      },
+      verb: {
+        display: {
+          'en-US': 'interacted',
+        },
+        id: VerbDefinition.interacted,
+      },
+    };
+
+    this.send(data);
+  }
+
+  private send(data: DataPayload) {
```

Comment:
Following this morning's discussion about analytics tools, it would be nice to be able to declare several listeners on these calls and make it modular. For example, someone could want to send these events to Google Analytics, Mixpanel or whatever.
The idea is to replace calls to this method by emitting an event. Then declaring event listener that subscribe to these events and do the actual logging to a given service. WDYT?

Reply:
This should be done in the backend, not here. Here the xAPI statement is incomplete, only the backend now, in this implementation, how to enrich it.